doc: clarify release guide (#3533)
openservicemesh/osm: docs/release_guide.md

  - [Update docs site](#update-docs-site)
  - [Announce the new release](#announce-the-new-release)
  - [Make version changes on main branch](#make-version-changes-on-main-branch)

>   - [Make version changes on docs.openservicemesh.io](#make-version-changes-on-docsopenservicemeshio)

Reviewer: Changes need to be made in the docs repo (osm-docs), please could you add a step for that and the relevant files if possible

$ git checkout -b release-<version> <commit-id> # ex: git checkout -b release-v0.4 0d05587
```

Push the release branch to the upstream repo (not forked), identified here by the `upstream` remote.
```console
$ git push upstream release-<version> # ex: git push upstream release-v0.4
```

> _Note_: Do not create a new branch for RCs in the upstream repo.

Author: I get you point. The emphasis of this line is "in the upstream repo".

> _Note_: Do not create a new branch for RCs in the upstream repo.

Author: Line removed.

## Update release branch with patches and versioning changes

> Create a new branch off of the release branch to maintain updates specific to the new version. Let's call it the patch branch. The patch branch should not be created in the upstream repo.

Reviewer 2: > The patch branch should not be created in the upstream repo.

If we delete the branch after its PR is merged, I don't see why it makes a difference where the branch lives. I understand every branch is currently captured by the upstream repo's branch protection rules which means an admin has to delete the branch and that's cumbersome.

I think a better solution to that particular problem is to fix in the branch protection rules vs. qualifying every current and future instruction about pushing short-lived branches with this same statement. Keeping this seems fine to me until then.

$ git push upstream release-<version> # ex: git push upstream release-v0.4
```

> Note that, DO NOT create a new branch for RCs in the upstream repo.

Reviewer: ```suggestion
_Note:_ Do not create a new branch for RCs in the upstream repo.
```

upload the packaged binaries and checksums as release assets, build and push Docker images for OSM and the demo to the
[`openservicemesh` organization](https://hub.docker.com/u/openservicemesh) on Docker Hub, and publish the Helm chart to the repo hosted at https://openservicemesh.github.io/osm.

To summarize, here is a diagram demonstrate the git branching during the release process.

> ![](../img/osm-git-release.jpg)

Reviewer 2: Personally, I find this diagram a bit overwhelming and hard to follow. This seems like a symptom that something in the text could be made more clear, in which case I'd rather clarify the text than supplement it with a diagram in the spirit of trying to keep this document as lean and easy to follow as we reasonably can.

I'm certainly open to keeping it if others find something like this helpful, but it does need at least one change:

> 3. merge patch to release branch

At this point, there have been no changes added to the patch branch. The changes from steps 4 and 5 can be made on the patch branch then a single PR can be made from the patch branch to the release branch.

Author: Discussed offline. Diagram removed for simplicity.

Reviewer 2: Certainly not necessary to handle in this PR, but another thought I had was to add a diagram that shows only the end result of the commits/branches/tags, since that's what's most important. Exactly how that result is achieved (PRs from certain branches, cherry-picking, merging) isn't all that important and may change, but it's good to document at least one method documented with concrete steps. I don't think the equivalent of that is described in words anywhere and would be useful context to have while performing the steps.

Again, I don't mean to dictate for everybody what is or isn't helpful, but I think it's important to avoid ambiguity in this doc as much as possible.

@allenlsy could you at least add the diagram back in a comment on this thread so we can get some more feedback?

Skip this step if the release is a release candidate (RC).

Open a pull request against the `main` branch making the same version updates as [above](#update-release-branch-with-patches-and-versioning-changes) so the latest release assets are referenced there.

## Make version changes on docs.openservicemesh.io

**Note**: do not perform this step for pre-releases.

Associated version numbers need to be updated in the [docs.openservicemesh.io](https://github.com/openservicemesh/osm-docs/) repo.

The files that need to be updated are:

> * [Control Plane Health Probes](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/control_plane_health_probes.md)

Reviewer 2: References to versions here are just sample output. I'd omit this file.

Skip this step if the release is a release candidate (RC).

Open a pull request against the `main` branch making the same version updates as [above](#update-release-branch-with-patches-and-versioning-changes) so the latest release assets are referenced there.

> ## Make version changes on docs.openservicemesh.io

Reviewer 2: We should add a note NOT to do this step for pre-releases (which includes RCs).

Associated version numbers need to be updated in the [docs.openservicemesh.io](https://github.com/openservicemesh/osm-docs/) repo.

> The files needs to be updated are:

Reviewer 3: nit:

```suggestion
The files that need to be updated are:
```

> The files that need to be updated are:

Reviewer 2: I only looked at the first few files here, but several seemed like they don't need to be included. Please take another look at these to trim it down as much as possible.

Author: I further trimmed the list down. The criteria is, we only update the docs that version number affects sample shell commands, or links to the latest code.

Reviewer 2: Thanks for doing that, I think some should be added back though. I'm going through https://github.com/openservicemesh/osm-docs/pull/109 to see if there's anything missing there. That PR changed a few spots that we don't need to update every time, but I'll leave comments on each of those spots. Then let's make this list based off of the changes there. I'll let you know when I'm finished reviewing that PR.

Sorry this is turning into a bigger deal than it probably should, but making this list complete and concise should save us headaches down the road.

Author: @nojnhuh I think osm-docs #109 is finished. I will merge this PR for v0.9.x release

Reviewer 4: Is it worth including a list here? Or should we guide devs to just use the search function?

Author: I actually think so. We do the same, and the list is not always accurate.

Author: @nojnhuh I removed the concrete files list here.

> * [Readme](https://github.com/openservicemesh/osm-docs/blob/main/README.md)

Reviewer 2: What exactly needs to be updated here? I only see references to release branches as an example, but those don't need to stay up to date.

* [Manual Demo](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/getting_started/manual_demo.md)

> * [Health Probes](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/application_health_probes.md)

Reviewer 2: The only references to versions I see here are in logs, so I don't think we need to update those every time we cut a release.

* [Control Plane Health Probes](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/control_plane_health_probes.md)
* [OSM MeshConfig](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/osm_mesh_config.md)

> * [Upgrade Guide](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/upgrade_guide.md)

Reviewer 2: I only see versions referenced here as an example.

> * [Installation](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/install/_index.md)

Reviewer 2: Where is there a version here at all?

Author: It's the link of OSM Chart:

https://github.com/openservicemesh/osm-docs/blob/main/content/docs/install/_index.md#using-the-helm-cli

Reviewer 2: Ah thanks, I see it now.

Skip this step if the release is a release candidate (RC).

Open a pull request against the `main` branch making the same version updates as [above](#update-release-branch-with-patches-and-versioning-changes) so the latest release assets are referenced there.

## Make version changes on docs.openservicemesh.io

**Note**: do not perform this step for pre-releases.

Associated version numbers need to be updated in the [docs.openservicemesh.io](https://github.com/openservicemesh/osm-docs/) repo.

> The suggested way is to search for hard coded current version number in the osm-docs repo, then update the ones that are will brake users' experience with if they follow the documentation, such as demonstration commands, reference links and anything that is strongly related to the next release. We don't need to update the version numbers that just serve the purpose of examples.

Reviewer 2: - Some references to code on release branches should be changed, so add a mention to also search for the release branch name
- grammar nits
- add a link to https://github.com/openservicemesh/osm-docs/pull/109 as an example.

```suggestion
The suggested way is to search for hard coded instances of the current version number and release branch name in the osm-docs repo, then update the ones that may break users' experiences if they follow the documentation, such as demonstration commands, reference links and anything that is strongly related to the next release. We don't need to update the version numbers that just serve the purpose of examples.

See https://github.com/openservicemesh/osm-docs/pull/109 for an example of the update from v0.8.4 to v0.9.0.
```
